src/base/ArrayUtils.i.hh

```diff
@@ -60,18 +60,57 @@ CELER_FUNCTION void normalize_direction(array<real_type, 3>* direction)
 
 //---------------------------------------------------------------------------//
 /*!
- * Rotate the direction about the given Z-based polar coordinates.
+ * Calculate a Cartesian vector from spherical coordinates.
  *
- * Note that rotation has loss of precision near the z axis. If this is a
- * problem we should consider using quaternions or another method robust for
- * rotations.
+ * Theta is the angle between the Z axis and the outgoing vector, and phi is
+ * the angle between the x axis and the projection of the vector onto the x-y
+ * plane.
  */
-inline CELER_FUNCTION void
-rotate_polar(real_type costheta, real_type phi, array<real_type, 3>* dir)
+inline CELER_FUNCTION Real3 from_spherical(real_type costheta, real_type phi)
 {
-    REQUIRE(is_soft_unit_vector(*dir, SoftEqual<real_type>(1e-6)));
     REQUIRE(costheta >= -1 && costheta <= 1);
 
+    const real_type sintheta = std::sqrt(1 - costheta * costheta);
+    return {sintheta * std::cos(phi), sintheta * std::sin(phi), costheta};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Rotate the direction about the given Z-based scatter direction.
+ *
+ * The equivalent to Shift's \code
+ * void cartesian_vector_transform(
+    double      costheta,
+    double      phi,
+    Vector_View vector)
+ * \endcode
+ * is the call
+ * \code
+   vector = rotate(from_spherical(costheta, phi), vector);
+ * \endcode
+ *
+ * This code effectively decomposes the given rotation vector \c rot into two
+ * Eulerian transforms, one with an angle \em theta about the \em y axis and
+ * one about \em phi rotating around the \em z axis. These two angles are the
+ * spherical coordinate transform of the given \c rot cartesian direction
+ * vector.
+ *
+ * There is some extra code in here to deal with loss of precision when the
+ * incident direction is along the \em z axis. As \c rot approaches \em z, the
+ * azimuthal angle \em phi must be calculated carefully from both the x and y
+ * components of the vector, not independently. If \c rot actually equals \em z
+ * then the azimuthal angle is completely indeterminate so we arbitrarily
+ * choose \c phi = 0.
+ *
+ * This function is often used for calculating exiting scattering angles. In
+ * that case, \c dir is the exiting angle from the scattering calculation, and
+ * \c rot is the original direction of the particle.
```

Comment:
Probably overkill, since you have complete documentation here, but I would add for complete reference
```cpp
* The direction vectors are defined 
* \f[
       \Omega = \sin\theta\cos\phi\mathbf{I} + \sin\theta\sin\phi\mathbf{j} + \cos\theta\mathbf{k}
* \f]
```